### PR TITLE
NMS-15431: Create a poller using Apache Apache mina-sshd java client

### PR DESCRIPTION
**Pull request by roskens**
### All Contributors

* [X] Have you read our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [X] Have you (electronically) signed the [OpenNMS Contributor Agreement](https://cla-assistant.io/OpenNMS/opennms)?

### Contribution Checklist

* Please [make an issue in the OpenNMS issue tracker](https://issues.opennms.org) if there isn't one already.<br />Once there is an issue, please:
  1. update the title of this PR to be in the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
  4. once you've created this PR, please link to it in a comment in the JIRA issue
  Don't worry if this sounds like a lot, we can help you get things set up properly.
* **If this code is likely to affect the UI, did you name your branch with `-smoke` in it to trigger smoke tests?**
* If this is a new or updated feature, is there documentation for the new behavior?
* If this is new code, are there unit and/or integration tests?
* If this PR targets a `foundation-*` branch, does it try to avoid changing files in `$OPENNMS_HOME/etc/`?

### What's Next?

A PR should be assigned at least 2 reviewers.  If you know that someone would be a good person to review your code, feel free to add them.

If you need help making additions or changes to the documentation related to your changes, please let us know.

In any case, if anything is unclear or you want help getting your PR ready for merge, please don't hesitate to say something in the comments here,
or in [the #opennms-development chat channel](https://chat.opennms.com/opennms/channels/opennms-development).

Once reviewer(s) accept the PR and the branch passes continuous integration, the PR is eligible for merge.

At that time, if you have commit access (are an OpenNMS Group employee or a member of the OGP) you are welcome to merge the PR when you're ready.
Otherwise, a reviewer can merge it for you.

Thanks for taking time to contribute!

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-15431

### CODE REVIEW

**Comment by dino2gnt** [comment]
Any docs or config examples?

**Comment by roskens** [comment]
> Any docs or config examples?

Yes, i've added those in. You'll to need to provide a username and a password or identity-file which it will use to authenticate with.

You can also provide it with a command to run, and if there are no failures, the service poll is assumed to be successful.

I've got ideas to enhance it with:
* server fingerprint verification - right now its using an accept-all host key verifier, so you could have it verify a specific host-key is presented.
* remote command output regex - verify the output of the remote command contains a pattern
* crypto: cipher's / mac's - only use specific ciphers when connecting, verify specific cipher's/mac's are returned.

**Comment by roskens** [comment]
> @roskens I _think_ I've fixed issues with building against non-internal PRs. Try rebasing this and giving it another shot. If you have problems I'll try and fix them so you can move on this.

Cool. I've rebased against tag opennms-31.0.6-1 and have pushed to my repo.

**Comment by roskens** [comment]
Conflict between sshd-common,sshd-core and sshd-osgi?

**Comment by roskens** [comment]
> There seems to be some additional karaf bits needed for this, since the admin shell isn't available after I start up opennms:
> 
> ```
> 2023-04-14T13:54:16,975 | WARN  | activator-1-thread-1 | Activator                        | 64 - org.apache.karaf.shell.ssh - 4.3.6 | Error starting activator
> java.lang.ClassCastException: Cannot cast org.apache.sshd.common.util.security.eddsa.EdDSASecurityProviderRegistrar to org.apache.sshd.common.util.security.SecurityProviderRegistrar
>         at java.lang.Class.cast(Class.java:3605) ~[?:?]
>         at org.apache.sshd.common.util.ReflectionUtils.newInstance(ReflectionUtils.java:74) ~[!/:?]
>         at org.apache.sshd.common.util.threads.ThreadUtils.createDefaultInstance(ThreadUtils.java:104) ~[!/:2.8.0]
>         at org.apache.sshd.common.util.threads.ThreadUtils.createDefaultInstance(ThreadUtils.java:112) ~[!/:2.8.0]
>         at org.apache.sshd.common.util.threads.ThreadUtils.createDefaultInstance(ThreadUtils.java:97) ~[!/:2.8.0]
>         at org.apache.sshd.common.util.security.SecurityUtils.register(SecurityUtils.java:430) ~[!/:2.8.0]
>         at org.apache.sshd.common.util.security.SecurityUtils.isBouncyCastleRegistered(SecurityUtils.java:395) ~[!/:2.8.0]
>         at org.apache.sshd.common.util.security.SecurityUtils.getRandomFactory(SecurityUtils.java:548) ~[!/:2.8.0]
>         at org.apache.sshd.common.BaseBuilder.fillWithDefaultValues(BaseBuilder.java:166) ~[!/:?]
>         at org.apache.sshd.server.ServerBuilder.fillWithDefaultValues(ServerBuilder.java:108) ~[!/:2.8.0]
>         at org.apache.sshd.server.ServerBuilder.fillWithDefaultValues(ServerBuilder.java:58) ~[!/:2.8.0]
>         at org.apache.sshd.common.BaseBuilder.build(BaseBuilder.java:273) ~[!/:?]
>         at org.apache.sshd.server.ServerBuilder.build(ServerBuilder.java:151) ~[!/:2.8.0]
>         at org.apache.sshd.server.ServerBuilder.build(ServerBuilder.java:58) ~[!/:2.8.0]
>         at org.apache.sshd.common.BaseBuilder.build(BaseBuilder.java:297) ~[!/:?]
>         at org.apache.sshd.server.SshServer.setUpDefaultServer(SshServer.java:442) ~[!/:2.8.0]
>         at org.apache.karaf.shell.ssh.Activator.createSshServer(Activator.java:183) ~[!/:?]
>         at org.apache.karaf.shell.ssh.Activator.createAndRunSshServer(Activator.java:121) ~[!/:?]
>         at org.apache.karaf.shell.ssh.Activator.doStart(Activator.java:116) ~[!/:?]
>         at org.apache.karaf.util.tracker.BaseActivator.run(BaseActivator.java:312) [!/:?]
>         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
>         at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
>         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
>         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
>         at java.lang.Thread.run(Thread.java:829) [?:?]
> ```

So this is due to apache mina sshd jars existing on the system class path (in $OPENMS_HOME/lib) instead of just inside $OPENNMS_HOME/system/.  See NMS-12788 for reference on how it affected bouncycastle.

**Comment by roskens** [comment]
Features this monitor does not include:

* custom SSH Banner verification
* use specific client MACs/ciphers 
* checks for what MACs/ciphers the remote server supports. (does it support A, but not B, etc.)
* server ssh host key acceptance & verification, ex: known hosts

File: smoke-test/src/test/java/org/opennms/smoketest/minion/MonitorsListCommandIT.java [hunk 1]
```diff
@@ -75,6 +75,7 @@ public class MonitorsListCommandIT {
             "org.opennms.netmgt.poller.monitors.TcpMonitor",
             "org.opennms.netmgt.poller.monitors.MemcachedMonitor",
             "org.opennms.netmgt.poller.monitors.ImapMonitor",
+            "org.opennms.netmgt.poller.monitors.MinaSshMonitor",
```
Comment (j-white) [inline smoke-test/src/test/java/org/opennms/smoketest/minion/MonitorsListCommandIT.java:78]: 👍

**Comment by RangerRick** [comment]
@roskens FYI, haven't forgotten about you, I'm investigating another SSHD-related weirdness now, and it will probably be useful to have that merged before we take a final crack at this

File: docs/modules/reference/pages/service-assurance/monitors/MinaSshMonitor.adoc [hunk 2]
```diff
@@ -0,0 +1,75 @@
+
+= MinaSshMonitor
+
+The MinaSshMonitor tests the availability of an SSH service.
+During the poll, it attempts to connect on the specified port and then authenticate in.
+If the authentication is successful, then the service is considered up.
+Optionally, a remote command to execute can be provided and if that is successful, the service is considered up.
+
```
Comment (mmahacek) [inline docs/modules/reference/pages/service-assurance/monitors/MinaSshMonitor.adoc:8]: Add a short statement to make it easy to see the difference between this and the standard Ssh Monitor.
```suggestion

== MinaSshMonitor vs SshMonitor

The MinaSshMonitor tests connecting and authenticating to a node via SSH.
The SshMonitor only tests that the SSH port is accepting connections.

```

On the SshMonitor.adoc page, add a similar statement in reverse order:
```
== SshMonitor vs MinaSshMonitor 

The SshMonitor only tests that the SSH port is accepting connections.
The MinaSshMonitor can be used to test both connecting and authenticating to a node via SSH.

**Review by mmahacek — COMMENTED** [review]
I can't speak to the code portions, but I have a suggestion on the docs side.

**Comment by RangerRick** [comment]
@mmahacek I incorporated your suggestion over in that PR.

**Pull request closed without merging**